services/appflowy-collaborate/src/group/group_init.rs

```diff
@@ -285,6 +300,89 @@ impl EditState {
   }
 }
 
+struct CollabUpdateStreamingImpl {
+  workspace_id: String,
+  object_id: String,
+  collab_redis_stream: Arc<CollabRedisStream>,
+  mem_cache: Mutex<Vec<Vec<u8>>>,
+  update_stream: Mutex<Option<StreamGroup>>,
+}
+
+impl CollabUpdateStreamingImpl {
+  fn new(workspace_id: &str, object_id: &str, collab_redis_stream: Arc<CollabRedisStream>) -> Self {
+    Self {
+      workspace_id: workspace_id.to_string(),
+      object_id: object_id.to_string(),
+      collab_redis_stream,
+      mem_cache: Default::default(),
+      update_stream: Default::default(),
+    }
+  }
+
+  /// Initialize redis update stream
+  /// After initialize the redis update stream, all pending updates will be sent immediately.
+  async fn init_update_stream(&self) -> Result<(), StreamError> {
+    let mut redis_update_stream = self
+      .collab_redis_stream
+      .collab_update_stream(
+        &self.workspace_id,
+        &self.object_id,
+        "collaborate_update_producer",
+      )
+      .await?;
+
+    // Send all pending updates
+    send_pending_updates(&mut redis_update_stream, &self.mem_cache).await;
+    *self.update_stream.lock().await = Some(redis_update_stream);
+    Ok(())
+  }
+}
+
+#[async_trait]
+impl CollabUpdateStreaming for CollabUpdateStreamingImpl {
+  async fn send_update(&self, update: Vec<u8>) {
+    if let Some(redis_update_stream) = self.update_stream.lock().await.deref_mut() {
+      if let Err(err) = redis_update_stream
+        .insert_message(CollabUpdateEvent::UpdateV1 {
+          encode_update: update,
+        })
+        .await
+      {
+        //FIXME: If fail to insert update to redis stream, which will cause missing update issue.
+        error!(
```

Comment:
Just like described above: if you use queues, you can make queue consumer task inside of collab update streaming. That task can return a `Result` which will propagate the error to `CollabUpdateStreaming` controller object (similar to how HTTP `server.await` works in main).

I propose this approach since the `CollabUpdateStreaming` has no control over itself, so it's not able to properly  restart on its own. Therefore critical failures needs to be propagated upwards.